File: ROS2/src/controller/hardware/RobotController.cpp
#include "chessbot/RobotController.hpp"

RobotController::RobotController(int numberOfAxes, std::string simulatorIp){
        this->numberOfAxes = numberOfAxes;
        this->zmqClient = new ZMQClient(simulatorIp);
}

RobotController::RobotController(int numberOfAxes, const char* simulatorIp){
        this->numberOfAxes = numberOfAxes;
        this->zmqClient = new ZMQClient(simulatorIp);
}

// -------------------------------- JOINTS --------------------------------------------------------

bool RobotController::SetJointPosition(std::vector<double> positions){
    if((int) positions.size() != numberOfAxes){
        std::cout << "Target position axis count mismatch:\n\tThe target position axis count: " << sizeof(positions)/sizeof(double) << " and the number of axes: " << numberOfAxes << " do not match." << std::endl;
        return false;
    }
    
    // Define message
    chessbot_simulator_control::Request_MessageType messageType;
    chessbot_simulator_control::Request req;
    req.MessageType_Parse("SET_JOINT_POSITION", &messageType);
    req.set_type(messageType);

    for(int i = 0; i < numberOfAxes; i++){ req.mutable_set_joint_position()->add_position(positions[i]); }

    // Send message
    std::string data;
    req.SerializeToString(&data);
    //std::cout << "Sending joint position command..." << std::endl;
    zmqClient->send(data);

    // Receive reply
    std::string reply = zmqClient->recv();
    chessbot_simulator_control::Acknowledge ack;
    ack.ParseFromString(reply);
    //std::cout << "Response to position command: " << ack.acknowledge() << std::endl;
    return true;
    
}

bool RobotController::SetJointVelocity(std::vector<double> velocities){
    if((int) velocities.size() != numberOfAxes){
        std::cout << "Target velocity axis count mismatch:\n\tThe target velocity axis count: " << sizeof(velocities)/sizeof(double) << " and the number of axes: " << numberOfAxes << " do not match." << std::endl;
        return false;
    }
    
    // Define message
    chessbot_simulator_control::Request_MessageType messageType;
    chessbot_simulator_control::Request req;
    req.MessageType_Parse("SET_JOINT_VELOCITY", &messageType);
    req.set_type(messageType);

    for(int i = 0; i < numberOfAxes; i++){ req.mutable_set_joint_velocity()->add_velocity(velocities[i]); }

    // Send message
    std::string data;
    req.SerializeToString(&data);
    //std::cout << "Sending joint velocity command..." << std::endl;
    zmqClient->send(data);

    // Receive reply
    std::string reply = zmqClient->recv();
    chessbot_simulator_control::Acknowledge ack;
    ack.ParseFromString(reply);
    //std::cout << "Response to velocity command: " << ack.acknowledge() << std::endl;
    return true;
    
}

bool RobotController::SetJointTorque(std::vector<double> torques){
    if((int)torques.size() != numberOfAxes){
        std::cout << "Target torque axis count mismatch:\n\tThe target torque axis count: " << sizeof(torques)/sizeof(double) << " and the number of axes: " << numberOfAxes << " do not match." << std::endl;
        return false;
    }
    
    // Define message
    chessbot_simulator_control::Request_MessageType messageType;
    chessbot_simulator_control::Request req;
    req.MessageType_Parse("SET_JOINT_TORQUE", &messageType);
    req.set_type(messageType);

    for(int i = 0; i < numberOfAxes; i++){ req.mutable_set_joint_torque()->add_torque(torques[i]); }

    // Send message
    std::string data;
    req.SerializeToString(&data);
    //std::cout << "Sending joint torque command..." << std::endl;
    zmqClient->send(data);

    // Receive reply
    std::string reply = zmqClient->recv();
    chessbot_simulator_control::Acknowledge ack;
    ack.ParseFromString(reply);
    //std::cout << "Response to torque command: " << ack.acknowledge() << std::endl;
    return true;
    
}

bool RobotController::SetJointPositionByAxisId(int axisId, double position){
    // Define message
    chessbot_simulator_control::Request_MessageType messageType;
    chessbot_simulator_control::Request req;
    req.MessageType_Parse("SET_JOINT_POSITION_BY_AXIS_ID", &messageType);
    req.set_type(messageType);

    req.mutable_set_joint_position_by_axis_id()->set_axis_id(axisId);
    req.mutable_set_joint_position_by_axis_id()->set_position(position);

    // Send message
    std::string data;
    req.SerializeToString(&data);
    //std::cout << "Sending joint position command by axis id..." << std::endl;
    zmqClient->send(data);

    // Receive reply
    std::string reply = zmqClient->recv();
    chessbot_simulator_control::Acknowledge ack;
    ack.ParseFromString(reply);
    //std::cout << "Response to joint position command: " << ack.acknowledge() << std::endl;
    return true;
}

bool RobotController::SetJointVelocityByAxisId(int axisId, double velocity){
    // Define message
    chessbot_simulator_control::Request_MessageType messageType;
    chessbot_simulator_control::Request req;
    req.MessageType_Parse("SET_JOINT_VELOCITY_BY_AXIS_ID", &messageType);
    req.set_type(messageType);

    req.mutable_set_joint_velocity_by_axis_id()->set_axis_id(axisId);
    req.mutable_set_joint_velocity_by_axis_id()->set_velocity(velocity);

    // Send message
    std::string data;
    req.SerializeToString(&data);
    //std::cout << "Sending joint velocity command by axis id..." << std::endl;
    zmqClient->send(data);

    // Receive reply
    std::string reply = zmqClient->recv();
    chessbot_simulator_control::Acknowledge ack;
    ack.ParseFromString(reply);
    //std::cout << "Response to joint velocity command: " << ack.acknowledge() << std::endl;
    return true;
}
bool RobotController::SetJointTorqueByAxisId(int axisId, double torque){
    // Define message
    chessbot_simulator_control::Request_MessageType messageType;
    chessbot_simulator_control::Request req;
    req.MessageType_Parse("SET_JOINT_TORQUE_BY_AXIS_ID", &messageType);
    req.set_type(messageType);

    req.mutable_set_joint_torque_by_axis_id()->set_axis_id(axisId);
    req.mutable_set_joint_torque_by_axis_id()->set_torque(torque);

    // Send message
    std::string data;
    req.SerializeToString(&data);
    //std::cout << "Sending joint torque command by axis id..." << std::endl;
    zmqClient->send(data);

    // Receive reply
    std::string reply = zmqClient->recv();
    chessbot_simulator_control::Acknowledge ack;
    ack.ParseFromString(reply);
    //std::cout << "Response to joint torque command: " << ack.acknowledge() << std::endl;
    return true;
}

auto RobotController::GetJointState() -> chessbot_simulator_control::JointStates {
    if(numberOfAxes <= 0){
        std::cout << "Number of axes invalid: " << numberOfAxes << std::endl;
        throw std::out_of_range("Number of axes invalid.");
    }

    // Define message
    chessbot_simulator_control::Request_MessageType messageType;
    chessbot_simulator_control::Request req;
    req.MessageType_Parse("GET_JOINT_STATE", &messageType);
    req.set_type(messageType);

    // Send message
    std::string data;
    req.SerializeToString(&data);
    zmqClient->send(data);

    // Receive reply
    std::string reply = zmqClient->recv();
    chessbot_simulator_control::JointStates jointStates;
    jointStates.ParseFromString(reply);
    
    return jointStates;
}

auto RobotController::GetJointStateByAxisId(int axisId) -> chessbot_simulator_control::JointState {
    if(numberOfAxes <= 0){
        std::cout << "Number of axes invalid: " << numberOfAxes << std::endl;
        throw std::out_of_range("Number of axes invalid.");
    }

    // Define message
    chessbot_simulator_control::Request_MessageType messageType;
    chessbot_simulator_control::Request req;
    req.MessageType_Parse("GET_JOINT_STATE_BY_AXIS_ID", &messageType);
    req.set_type(messageType);

    req.mutable_get_joint_state_by_axis_id()->set_axis_id(axisId);

    // Send message
    std::string data;
    req.SerializeToString(&data);
    zmqClient->send(data);

    // Receive reply
    std::string reply = zmqClient->recv();
    chessbot_simulator_control::JointState jointState;
    jointState.ParseFromString(reply); 

    return jointState;
}

// -------------------------------- GRIPPER --------------------------------------------------------

bool RobotController::SetGripperPosition(double position){ 
    // Define message
    chessbot_simulator_control::Request_MessageType messageType;
    chessbot_simulator_control::Request req;
    req.MessageType_Parse("SET_GRIPPER_POSITION", &messageType);
    req.set_type(messageType);

    req.mutable_set_gripper_position()->set_position(position);

    // Send message
    std::string data;
    req.SerializeToString(&data);
    //std::cout << "Sending gripper position command..." << std::endl;
    zmqClient->send(data);

    // Receive reply
    std::string reply = zmqClient->recv();
    chessbot_simulator_control::Acknowledge ack;
    ack.ParseFromString(reply);
    //std::cout << "Response to position command: " << ack.acknowledge() << std::endl;
    return true;
    
}

auto RobotController::GetGripperState() -> chessbot_simulator_control::GripperState {

    // Define message
    chessbot_simulator_control::Request_MessageType messageType;
    chessbot_simulator_control::Request req;
    req.MessageType_Parse("GET_GRIPPER_STATE", &messageType);
    req.set_type(messageType);

    // Send message
    std::string data;
    req.SerializeToString(&data);
    zmqClient->send(data);

    // Receive reply
    std::string reply = zmqClient->recv();
    chessbot_simulator_control::GripperState gripperState;
    gripperState.ParseFromString(reply); 

    return gripperState;
}

// -------------------------------- MISC --------------------------------------------------------


double RobotController::GetToleranceValue(){ return tolerance; }

RobotController::~RobotController(){
    delete zmqClient;
}

File: Webots/controller_and_interface/robot_controller.py
import sys
from controller import Robot
from concurrent import futures
import threading
import grpc
import ikpy
from ikpy import chain
import zmq
from enum import Enum
from time import perf_counter_ns
import simulator_control_pb2

DEG2RAD = 0.0174532925
RAD2DEG = 57.2957795

class bcolors:
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'

class zmq_time_sync_server():
    def __init__(self):
        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.PUB)
        self.socket.bind("tcp://*:5556")
    
    def publish(self):
        # Heartbeat message - has no meaning
        self.socket.send(b"%d %d" % (1, 1))

class zmq_server():
    def __init__(self):
        
        self.context = zmq.Context()

        self.socket = self.context.socket(zmq.REP)
        self.socket.bind("tcp://*:5555")
        
        # This dictionary stores which function to execute upon receiving a zmq request
        self.function_container = {
            simulator_control_pb2.Request.MessageType.GET_ROBOT_PARAMETER: robot_object.os_get_robot_parameters,

            ## JOINTS ##
            simulator_control_pb2.Request.MessageType.SET_JOINT_POSITION: robot_object.os_set_target_joint_position,
            simulator_control_pb2.Request.MessageType.SET_JOINT_VELOCITY: robot_object.os_set_target_joint_velocity,
            simulator_control_pb2.Request.MessageType.SET_JOINT_TORQUE: robot_object.os_set_target_joint_torque,
            simulator_control_pb2.Request.MessageType.SET_JOINT_POSITION_BY_AXIS_ID: robot_object.os_set_target_joint_position_by_axis_id,
            simulator_control_pb2.Request.MessageType.SET_JOINT_VELOCITY_BY_AXIS_ID: robot_object.os_set_target_joint_velocity_by_axis_id,
            simulator_control_pb2.Request.MessageType.SET_JOINT_TORQUE_BY_AXIS_ID: robot_object.os_set_target_joint_torque_by_axis_id,

            simulator_control_pb2.Request.MessageType.GET_JOINT_STATE: robot_object.os_get_joint_state,
            simulator_control_pb2.Request.MessageType.GET_JOINT_STATE_BY_AXIS_ID: robot_object.os_get_joint_state_by_axis_id,
            
            ## GRIPPER ##
            simulator_control_pb2.Request.MessageType.SET_GRIPPER_POSITION: robot_object.os_set_gripper_position,
            simulator_control_pb2.Request.MessageType.GET_GRIPPER_STATE: robot_object.os_get_gripper_state,
            }

        self.zmq_loop()


    def zmq_loop(self):
        while enable_thread:
            try: request_unparsed = self.socket.recv(zmq.NOBLOCK)
            except: continue
            request = simulator_control_pb2.Request()
            request.ParseFromString(request_unparsed)
            thread_lock.acquire()
            result = self.function_container[request.type](request=request)
            thread_lock.release()
            self.socket.send(result.SerializeToString())

class control_mode(Enum):
    POSITION_CONTROL = 1
    VELOCITY_CONTROL = 2
    TORQUE_CONTROL = 3

class Robot_Object():
    def __init__(self, urdf_name):
        self.urdf_name = urdf_name
        self.TIME_STEP = 16
        self.robot = Robot()
        self.positions = []
        self.velocities = []
        self.torques = []
        self.gripper_position = None
        self.machine_data_dictionary = {
            "maximum_positions":[],
            "minimum_positions":[],
            "maximum_velocities":[],
            "minimum_velocities":[],
            "maximum_efforts":[],
            "minimum_efforts":[]
            }
        self.motors = []
        self.sensors = []
        self.gripper_motors = []
        self.gripper_sensors = []

        # Using position control by default
        self.control_mode = control_mode.POSITION_CONTROL

        armChain = ikpy.chain.Chain.from_urdf_file('resource/' + self.urdf_name)
        
        # Arm setup
        for i in range(len(armChain.links)):
            if(armChain.links[i].name.startswith('Joint')):
                motor = self.robot.getDevice(armChain.links[i].name)
                motor.enableTorqueFeedback(self.TIME_STEP)
                position_sensor = self.robot.getDevice(armChain.links[i].name+ '_sensor')
                position_sensor.enable(self.TIME_STEP)
                self.motors.append(motor)
                self.sensors.append(position_sensor)
                self.machine_data_dictionary["maximum_positions"].append(motor.getMaxPosition())
                self.machine_data_dictionary["minimum_positions"].append(motor.getMinPosition())
                self.machine_data_dictionary["maximum_velocities"].append(motor.getMaxVelocity())
                self.machine_data_dictionary["minimum_velocities"].append(0.0)
                self.machine_data_dictionary["maximum_efforts"].append(motor.getMaxTorque())
                self.machine_data_dictionary["minimum_efforts"].append(0.0)

        # Gripper setup
        gripper_joints = ['finger_joint', 'left_inner_knuckle_joint', 'left_inner_finger_joint', 'right_inner_knuckle_joint', 'right_inner_finger_joint', 'right_outer_knuckle_joint']
        for joint in gripper_joints:
            gripper_motor = self.robot.getDevice(joint)
            gripper_sensor = self.robot.getDevice(joint +  '_sensor')
            gripper_sensor.enable(self.TIME_STEP)
            self.gripper_motors.append(gripper_motor)
            self.gripper_sensors.append(gripper_sensor)
        
        # Default values
        self.axes_number = len(self.motors)

        self.target_positions = [None] * self.axes_number
        self.target_velocities = [None] * self.axes_number
        self.target_torques = [None] * self.axes_number
        self.target_gripper_position = None


## --------------------- JOINT FUNCTIONS -----------------------------------------
    def os_set_target_joint_position(self, request):
        #print("Received joint position target:")
        self.target_positions = [request.set_joint_position.position[i] for i in range(self.axes_number)]
        #print(self.target_positions)
        response = simulator_control_pb2.Acknowledge()
        response.acknowledge = "OK"
        self.control_mode = control_mode.POSITION_CONTROL
        return response

    def os_set_target_joint_velocity(self, request):
        #print("Received joint velocity target:")
        self.target_velocities = [request.set_joint_velocity.velocity[i] for i in range(self.axes_number)]
        #print(self.target_velocities)
        response = simulator_control_pb2.Acknowledge()
        response.acknowledge = "OK"
        self.control_mode = control_mode.VELOCITY_CONTROL
        return response

    def os_set_target_joint_torque(self, request):
        #print("Received joint torque target:")
        self.target_torques = [request.set_joint_torque.torque[i] for i in range(self.axes_number)]
        #print(self.target_torques)
        response = simulator_control_pb2.Acknowledge()
        response.acknowledge = "OK"
        self.control_mode = control_mode.TORQUE_CONTROL
        return response
    
    def os_set_target_joint_position_by_axis_id(self, request):
        #print("Received joint position target:")
        axis_id = request.set_joint_position_by_axis_id.axis_id
        self.target_positions[axis_id]=request.set_joint_position_by_axis_id.position
        #print(str(axis_id) + "\t" + str(self.target_positions[axis_id]))
        response = simulator_control_pb2.Acknowledge()
        response.acknowledge = "OK"
        self.control_mode = control_mode.POSITION_CONTROL
        return response

    def os_set_target_joint_velocity_by_axis_id(self, request):
        #print("Received joint velocity target:")
        axis_id = request.set_joint_position_by_axis_id.axis_id
        self.target_velocities[axis_id]=request.set_joint_velocity_by_axis_id.velocity
        #print(str(axis_id) + "\t" + str(self.target_velocities[axis_id]))
        response = simulator_control_pb2.Acknowledge()
        response.acknowledge = "OK"
        self.control_mode = control_mode.VELOCITY_CONTROL
        return response

    def os_set_target_joint_torque_by_axis_id(self, request):
        #print("Received joint torque target:")
        axis_id = request.set_joint_position_by_axis_id.axis_id
        self.target_torques[axis_id]=request.set_joint_torque_by_axis_id.torque
        #print(str(axis_id) + "\t" + str(self.target_torques[axis_id]))
        response = simulator_control_pb2.Acknowledge()
        response.acknowledge = "OK"
        self.control_mode = control_mode.TORQUE_CONTROL
        return response
    
    def os_get_joint_state(self, request):
        response = simulator_control_pb2.JointStates()
        for i in range(self.axes_number):
            joint_state = response.states.add()
            joint_state.axis_id = i
            joint_state.position = self.positions[i]
            joint_state.velocity = self.velocities[i]
            joint_state.torque = self.torques[i]
        return response
    
    def os_get_joint_state_by_axis_id(self, request):
        response = simulator_control_pb2.JointState()
        axis_id = request.get_joint_state_by_axis_id.axis_id
        response.axis_id = axis_id
        response.position = self.positions[axis_id]
        response.velocity = self.velocities[axis_id]
        response.torque = self.torques[axis_id]
        return response

## ------------------------ GRIPPER --------------------------------------------
    def os_set_gripper_position(self, request):
        #print("Received gripper position target:")
        # Set the gripper joint position
        self.target_gripper_position = request.set_gripper_position.position
        response = simulator_control_pb2.Acknowledge()
        response.acknowledge = "OK"
        return response

    def os_get_gripper_state(self, request):
        response = simulator_control_pb2.GripperState()
        response.position = self.gripper_position
        return response

    
## --------------------- MISCELLANEOUS -----------------------------------------

    def os_get_robot_parameters(self, request):
        response = simulator_control_pb2.RobotParameter()
        
        response.axes_number = self.axes_number

        for i in range(self.axes_number):
            position_limit = simulator_control_pb2.JointPositionLimit()
            position_limit.upper_limit = self.machine_data_dictionary["maximum_positions"][i]
            position_limit.lower_limit = self.machine_data_dictionary["minimum_positions"][i]

            velocity_limit = simulator_control_pb2.JointVelocityLimit()
            velocity_limit.upper_limit = self.machine_data_dictionary["maximum_velocities"][i]
            velocity_limit.lower_limit = self.machine_data_dictionary["minimum_velocities"][i]

            effort_limit = simulator_control_pb2.JointEffortLimit()
            effort_limit.upper_limit = self.machine_data_dictionary["maximum_efforts"][i]
            effort_limit.lower_limit = self.machine_data_dictionary["minimum_efforts"][i]

            joint_limit = simulator_control_pb2.JointLimit()

            joint_limit.type = simulator_control_pb2.JointLimit.JointType.ROTATIONAL
            joint_limit.position_limit.CopyFrom(position_limit)
            joint_limit.velocity_limit.CopyFrom(velocity_limit)
            joint_limit.effort_limit.CopyFrom(effort_limit)

            response.joint_limit.extend([joint_limit])

        return response
        

thread_lock = threading.Lock() # for multithreading
robot_object = Robot_Object(sys.argv[1])
enable_thread = True

def main():

    zmq_thread = threading.Thread(target = zmq_server)
    zmq_thread.start()


    zmq_heartbeat = zmq_time_sync_server()

    print(f"{bcolors.OKGREEN}ALL SERVERS ARE READY - STARTING SIMULATION{bcolors.ENDC}")


    while robot_object.robot.step(robot_object.TIME_STEP) != -1:

        thread_lock.acquire()
        robot_object.positions = [robot_object.sensors[i].getValue() for i in range(len(robot_object.motors))]
        robot_object.velocities = [robot_object.motors[i].getVelocity() for i in range(len(robot_object.motors))]
        robot_object.torques = [robot_object.motors[i].getTorqueFeedback() for i in range(len(robot_object.motors))]
        robot_object.gripper_position = robot_object.gripper_sensors[0].getValue()

        if(robot_object.control_mode is control_mode.POSITION_CONTROL):
            [robot_object.motors[i].setPosition(robot_object.target_positions[i]) if robot_object.target_positions[i] is not None else None for i in range(len(robot_object.motors))]
        if(robot_object.control_mode is control_mode.VELOCITY_CONTROL):
            for i in range(len(robot_object.motors)):
                robot_object.motors[i].setPosition(float('+inf'))
                robot_object.motors[i].setVelocity(robot_object.target_velocities[i]) if robot_object.target_velocities[i] is not None else None
        if(robot_object.control_mode is control_mode.TORQUE_CONTROL):
            [robot_object.motors[i].setTorque(robot_object.target_torques[i]) if robot_object.target_torques[i] is not None else None for i in range(len(robot_object.motors))]
            
        [robot_object.gripper_motors[i].setPosition(robot_object.target_gripper_position) if robot_object.target_gripper_position is not None else None for i in range(len(robot_object.gripper_motors))]

        thread_lock.release()

        zmq_heartbeat.publish()
    

if __name__ == '__main__':
    main()
    thread_lock.acquire()
    enable_thread = False
    thread_lock.release()
